kernel/comps/virtio/src/device/socket/buffer.rs

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{collections::LinkedList, sync::Arc};
+
+use align_ext::AlignExt;
+use aster_frame::{
+    sync::SpinLock,
+    vm::{Daddr, DmaDirection, DmaStream, HasDaddr, VmAllocOptions, VmReader, VmWriter, PAGE_SIZE},
+};
+use aster_network::dma_pool::{DmaPool, DmaSegment};
+use pod::Pod;
+use spin::Once;
+
+pub struct TxBuffer {
+    dma_stream: DmaStream,
+    nbytes: usize,
+}
+
+impl TxBuffer {
+    pub fn new<H: Pod>(header: &H, packet: &[u8]) -> Self {
+        let header = header.as_bytes();
+        let nbytes = header.len() + packet.len();
+
+        let dma_stream = if let Some(stream) = get_tx_stream_from_pool(nbytes) {
+            stream
+        } else {
+            let segment = {
+                let nframes = (nbytes.align_up(PAGE_SIZE)) / PAGE_SIZE;
+                VmAllocOptions::new(nframes).alloc_contiguous().unwrap()
+            };
+            DmaStream::map(segment, DmaDirection::ToDevice, false).unwrap()
+        };
+
+        let mut writer = dma_stream.writer().unwrap();
+        writer.write(&mut VmReader::from(header));
+        writer.write(&mut VmReader::from(packet));
+
+        let tx_buffer = Self { dma_stream, nbytes };
+        tx_buffer.sync();
+        tx_buffer
+    }
+
+    pub fn writer(&self) -> VmWriter<'_> {
+        self.dma_stream.writer().unwrap().limit(self.nbytes)
+    }
+
+    fn sync(&self) {
+        self.dma_stream.sync(0..self.nbytes).unwrap();
+    }
+
+    pub fn nbytes(&self) -> usize {
+        self.nbytes
+    }
+}
+
+impl HasDaddr for TxBuffer {
+    fn daddr(&self) -> Daddr {
+        self.dma_stream.daddr()
+    }
+}
+
+impl Drop for TxBuffer {
+    fn drop(&mut self) {
+        TX_BUFFER_POOL
+            .get()
+            .unwrap()
+            .lock_irq_disabled()
+            .push_back(self.dma_stream.clone());
+    }
+}
+
+pub struct RxBuffer {
+    segment: DmaSegment,
+    header_len: usize,
+    packet_len: usize,
+}
+
+impl RxBuffer {
+    pub fn new(header_len: usize) -> Self {
```

Comment:
Could you please try to reuse the `RxBuffer` code instead of defining a same struct?

Below is one option. Pool is added as a parameter.

```rust
/// `new` may be renamed as `pool_alloc` if `pool` is a parameter
pub fn pool_alloc(header_len: usize, pool: &DmaPool) -> Self;
```

Same modifications should be done on `TxBuffer`